checker/src/abstract_domains.rs

```diff
@@ -345,50 +345,28 @@ impl AbstractDomain {
                     }
                 }
             }
-            // x == 0 is the same as !x when x is Boolean variable. Canonicalize it to the latter.
+            // !x == 0 is the same as x when x is Boolean. Canonicalize it to the latter.
             (
-                Expression::Variable { var_type, .. },
+                Expression::Not { operand },
                 Expression::CompileTimeConstant(ConstantDomain::U128(val)),
             ) => {
-                if *var_type == ExpressionType::Bool && *val == 0 {
-                    return self.not();
+                if *val == 0 && operand.expression.infer_type() == ExpressionType::Bool {
```

Comment:
Heuristic simplification is remarkably effective and reasonably fast if properly k-limited. When driven by actually observed patterns, I think it makes sense to employ them. A full fledged SMT solver is obviously more complete but usually much slower than the heuristics (and when not, it is because the solver uses heuristics. :-). In MIRAI I'm trying to have the best of both worlds (and to throw in some Abstract Domain magic as well, at some point).